### PR TITLE
Implement fabric:open_revs/3

### PR DESCRIPTION
**Pull request by nickva**
`fabric:open_revs/3` is the batched counterpart to `fabric:open_revs/4`.
Instead of fetching revisions for only one document, like `fabric:open_revs/4`,
it can fetch batches of document revisions. The main optimization is achieved
by grouping documents by shard ranges and then spawning shard workers only once
for each range (essentially doing what fabric:update_docs/2,3 does).

The idea is to use this implementation for a more efficient version of _bulk_get.
So far this is just the fabric API, not plugged into the HTTP API yet.

The backend API for this is already available as implemented in
https://github.com/apache/couchdb/pull/4185.

While the implementation is intended to be as close as possible to
`fabric:open_revs/4` there a key difference -- since we're now dealing with
potentially large batches, we choose not to perform immediate read repair.
Instead, we let the internal replicator handle that. Read repair can spawn an
unbounded number of processes, and now with batches those could be large
fabric:update_docs/2,3 requests. Issuing lots of those, especially when a
cluster has a hard time keeping up, might not be a good idea.

For now we leave the `fabric:open_revs/4` implementation intact, as opposed to
re-wiring it to call `fabric:open_revs/3` with a one-element list. This helps
have a test reference point (see fabric_open_revs_test for an example), allows
for a configurable fallback for _bulk_get implementation, and lets individual doc
GET fetches work exactly as they have been.

A few details about the implementation itself:

  * The coordinator state is kept in the `#st{}` record. Each requested doc ID
and revisions list argument keeps track of its quorum logic and gathers
responses in its individual `#req{}` record. There is exactly one of those for
each requested `{{Id, Revs}, DocOpts}` argument.

  * Since the `{{Id, Revs}, DocOpts}` arguments are not guaranteed to be unique,
we have to keep track of them so we can return the results in the same order.
To help with that we tag each with a unique reference, and keep track of it in
`{atts_since, ...}` revisions, it helps to stash away the arguments separately
and just operating with reference tags. Those references are usually called
`ArgRef` and `ArgsRefs` in the code.

  * The `#st.workers` map keeps track of the spawned workers and which args
they are responsible for. As each worker returns, or is knocked out by an
error, it gets removed from the `#st.workers` map and at the same time all the
individual requests (`#req{}` record) are updated. After the update we check if
read quorum has been met and if it's still possible to succeed with the
remaining workers.

  * There is a 100% test coverage. The tests in the `fabric_open_revs_test`
module are more end-to-end and they exercise various options. The tests in
`fabric_open_revs` test how quorum logic and error handling works by emulating
various workers returning and how revisions are merged together into the final
result.

Issue: https://github.com/apache/couchdb/issues/4183

### CODE REVIEW

File: src/fabric/test/eunit/fabric_open_revs_test.erl [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    % Also check that workers were cleaned up properly
+    ?assertEqual(1, meck:num_calls(rexi, kill_all, 1)).
+
+open_revs_single(DbName, IdRevsOpts, Options) ->
+    {ok, [Res]} = fabric:open_revs(DbName, [IdRevsOpts], Options),
+    % Validate batched open_revs/3 call is equivalent to calling
+    % the single open_revs/4 with that same Id and Revs args.
+    {{Id, Revs}, DocOpts} = IdRevsOpts,
+    {ok, ResSingle} = fabric:open_revs(DbName, Id, Revs, DocOpts ++ Options),
+    ?assertEqual(ResSingle, Res),
```
Comment (jaydoane) [inline src/fabric/test/eunit/fabric_open_revs_test.erl:212]: This is a very nice touch, although I'm guessing once we're convinced they are equivalent, the single version can be eliminated in favor of the more general function?

**Review by jaydoane — APPROVED** [review]
I very much appreciate how this is implemented without changing the existing functionality since it makes it possible to easily compare the two side by side with the same code. Bravo!

Also impressive test coverage 🎉 
```
fabric_open_revs              :  99%
fabric_open_revs_test         :  98%
```
I wonder if there's a minor bug in the coverage plugin because when I look at e.g. `.eunit/fabric_open_revs_test.COVER.html` I don't actually see _any_ code that's not covered by tests.

File: src/fabric/src/fabric_open_revs.erl [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+        true ->
+            {ok, St#st{workers = Workers1, reqs = Reqs1}};
+        false ->
+            stop_workers(Workers1),
+            {stop, finalize(St#st.args, Reqs1)}
+    end;
+handle_message({rexi_DOWN, _, {_, NodeRef}, _}, _Worker, #st{} = St) ->
+    #st{workers = Workers, reqs = Reqs} = St,
+    FilterFun = fun(#shard{node = N}) -> N =:= NodeRef end,
+    DeadKeys = lists:filter(FilterFun, maps:keys(Workers)),
```
[the remark this answers is not in the document]
Comment (nickva) [reply src/fabric/src/fabric_open_revs.erl:68]: All the workers sent to that particular node, referenced by `NodeRef` which would be just the node name like `node1@address...`, would be dead. However, there is a chance that there are still other shard copies for the same range which may still return.

File: src/fabric/src/fabric_open_revs.erl [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+    RevTree = couch_key_tree:multi_merge([], Paths),
+    TreeLeafs = couch_key_tree:get_all_leafs(RevTree),
+    lists:map(fun path_to_reply/1, TreeLeafs).
+
+path_to_reply({?REV_MISSING, {Pos, [Rev]}}) ->
+    {{not_found, missing}, {Pos, Rev}};
+path_to_reply({#doc{} = Doc, _}) ->
+    {ok, Doc}.
+
+is_not_found([]) ->
```
[the remark this answers is not in the document]
Comment (nickva) [reply src/fabric/src/fabric_open_revs.erl:238]: Agree, that's a better name.

File: src/fabric/src/fabric_open_revs.erl [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+            Else
+    after
+        rexi_monitor:stop(RexiMon)
+    end.
+
+handle_message([_ | _] = Resps, Worker, #st{} = St) ->
+    #st{workers = Workers, reqs = Reqs, r = R} = St,
+    {ArgsRefs, Workers1} = maps:take(Worker, Workers),
+    ArgsResps = lists:zip(ArgsRefs, Resps),
+    Reqs1 = lists:foldl(fun responses_fold/2, Reqs, ArgsResps),
```
[the remark this answers is not in the document]
Comment (nickva) [reply src/fabric/src/fabric_open_revs.erl:57]: Since `reponses_fold/2` does de-duping but also other updates `Req1` could make sense there too.

File: src/fabric/src/fabric_open_revs.erl [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+    #req{rcnt = R, wcnt = W, responses = Resps} = Req,
+    % De-duplicate identical results as we go along
+    Resps1 = lists:umerge(Resps, lists:usort(NewResp)),
+    % If responses don't match or got a "not found", don't bump rcnt so we can
+    % wait for more workers.
+    NewR =
+        case {is_not_found(NewResp), length(Resps)} of
+            {true, _} -> R;
+            {_, 0} -> R + 1;
+            {_, OldLen} when OldLen == length(Resps1) -> R + 1;
```
[the remark this answers is not in the document]
Comment (nickva) [reply src/fabric/src/fabric_open_revs.erl]: I updated this logic a bit to extract the `OldLen` and `NewLen` out of the case statement. It's a tiny bit less efficient but more clear to read.

```erlang
 case {any_not_found(NewResp), OldLen} of
            {true, _} -> R;
            {_, 0} -> R + 1;
            {_, L} when L == NewLen -> R + 1;
            {_, L} when L < NewLen -> R
        end,
```

In this case `{_, 0}` would not be covered by` {_, L} when L == NewLen -> R + 1` as  when the first response comes `OldLen =0` but `NewLen` would be 1. So, they are different, however we do want to bump the `R` count because it's the only response we received and don't have a reason to consider that it's a diverging result.

As soon as we have more than one response (`OldLen = 1`) then we switch the logic and bump the `R` count only if we start getting the same length, that is we get the same response. If we get different responses (`OldLen < NewLen`) we stop bumping it because we're hoping to get a more worker responses.

File: src/fabric/src/fabric_open_revs.erl [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+t_no_quorum_on_different_responses(_) ->
+    % Got different revisions, so we're waiting for more workers.
+    S0 = #st{workers = Workers0} = st0(),
+    [W1, W2, W3] = lists:sort(maps:keys(Workers0)),
+    {ok, S1} = handle_message([[foo1(), bar1()]], W1, S0),
+    Res1 = handle_message([[foo2(), bar1()]], W2, S1),
+    ?assertMatch({ok, #st{}}, Res1),
+    {ok, S2} = Res1,
+    Res2 = handle_message([[foo1(), bar1()]], W3, S2),
+    ?assertEqual({stop, [[foo2(), bar1()]]}, Res2).
```
[the remark this answers is not in the document]
Comment (nickva) [reply src/fabric/src/fabric_open_revs.erl:390]: Good idea, I'll add a comment about the stopping condition.

File: src/fabric/src/fabric_open_revs.erl [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+t_ancestors_merge_correctly(_) ->
+    % Ancestors are merged as internal nodes correctly
+    S0 = #st{workers = Workers0} = st0(),
+    [W1, W2, W3] = lists:sort(maps:keys(Workers0)),
+    {ok, S1} = handle_message([[foo1()]], W1, S0),
+    {ok, S2} = handle_message([[foo2()]], W2, S1),
+    ?assertEqual({stop, [[foo2()]]}, handle_message([[]], W3, S2)).
+
+t_stemmed_merge_correctly(_) ->
+    % Ancestors are merged as internal nodes correctly
```
[the remark this answers is not in the document]
Comment (nickva) [reply src/fabric/src/fabric_open_revs.erl:426]: Yeah, it make sense there too. But I'll add a node that it include stemmed revisions too.

File: src/fabric/test/eunit/fabric_open_revs_test.erl [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+    Docs = [
+        #doc{id = <<"a">>, revs = {1, [<<"z">>]}},
+        #doc{id = <<"a">>, revs = {2, [<<"x">>, <<"z">>]}},
+        #doc{id = <<"a">>, revs = {2, [<<"y">>, <<"z">>]}}
+    ],
+    Opts = [?REPLICATED_CHANGES],
+    {ok, []} = fabric:update_docs(DbName, Docs, Opts),
+    {Ctx, DbName}.
+
+teardown({Ctx, _DbName}) ->
```
[the remark this answers is not in the document]
Comment (nickva) [reply src/fabric/test/eunit/fabric_open_revs_test.erl:64]: Oh good find. Yes, it should!

File: src/fabric/test/eunit/fabric_open_revs_test.erl (hunk 1 above)
Comment (nickva) [reply src/fabric/test/eunit/fabric_open_revs_test.erl:212]: At some point they could. But since we managed to keep both implementations around we're taking advantage of it testing various corner cases both for equivalence between the old and the new API.

Eventually the old version would just be re-implemented as `fabric:open_revs(Db, [{Id, Revs, DocOpts}], Options)` not unlike we already do for `update_doc` and `update_docs`

**Comment by nickva** [comment]
> I wonder if there's a minor bug in the coverage plugin because when I look at e.g. `.eunit/fabric_open_revs_test.COVER.html` I don't actually see _any_ code that's not covered by tests.

I noticed that too. The coverage was reported as 100% at some point. But then I did a minor refactoring and it switched 99% even though it still didn't show any red "missed" lines in the output. I suspect it's rounding error somewhere.
